### PR TITLE
Pass 'not' to commit count

### PR DESCRIPTION
**Pull request by mattwalo32**
Due to #24409 , we can now specify '--not' when getting all commits from a repo to exclude commits from a different branch.

When I wrote that PR, I forgot to also update the code that counts the number of commits in the repo. So now, if the --not option is used, it may return too many commits, which can indicate that another page of data is available when it is not.

This PR passes --not to the commands that count the number of commits in a repo

### CODE REVIEW

**Comment by wxiaoguang** [comment]
Some thoughts:

1. `CommitsCount(ctx context.Context, repoPath, not string, revision ...string)` seems very fragile, too many similar arguments , it would cause bugs sooner or later. 
    * Using some `Opt` struct to pass complex arguments could be an approach
2. It needs some tests, to avoid something unexpected happening again, and make the code could be maintained/improved in the future.

File: modules/git/commit.go [hunk 1]
```diff
@@ -177,13 +190,18 @@ func CommitsCountFiles(ctx context.Context, repoPath string, revision, relpath [
...
-func CommitsCount(ctx context.Context, repoPath string, revision ...string) (int64, error) {
-	return CommitsCountFiles(ctx, repoPath, revision, []string{})
+func CommitsCount(ctx context.Context, opts CommitsCountOptions) (int64, error) {
+	return CommitsCountFiles(ctx, opts)
 }
 
 // CommitsCount returns number of total commits of until current revision.
-func (c *Commit) CommitsCount() (int64, error) {
-	return CommitsCount(c.repo.Ctx, c.repo.Path, c.ID.String())
+func (c *Commit) CommitsCount(not string) (int64, error) {
```
Comment (wxiaoguang) [inline modules/git/commit.go]: There is only one call to `commit.CommitsCount(not)`

I think it better to keep the old `commit.CommitsCount()` (no `not` argument)

And make that only-one `commit.CommitsCount(not)` to use `CommitsCountOptions` directly. Then a lot of code deosn't need to change, there won't be many `CommitsCount("")` lines.

File: modules/git/commit.go [hunk 2]
```diff
@@ -177,13 +190,18 @@ func CommitsCountFiles(ctx context.Context, repoPath string, revision, relpath [
 }
 
 // CommitsCount returns number of total commits of until given revision.
-func CommitsCount(ctx context.Context, repoPath string, revision ...string) (int64, error) {
-	return CommitsCountFiles(ctx, repoPath, revision, []string{})
+func CommitsCount(ctx context.Context, opts CommitsCountOptions) (int64, error) {
+	return CommitsCountFiles(ctx, opts)
```
Comment (lunny) [inline modules/git/commit.go]: Looks like the two functions now could be merged as one?

File: modules/git/commit.go [hunk 3]
```diff
@@ -177,13 +190,18 @@ func CommitsCountFiles(ctx context.Context, repoPath string, revision, relpath [
...
 }
 
 // CommitsCount returns number of total commits of until current revision.
 func (c *Commit) CommitsCount() (int64, error) {
-	return CommitsCount(c.repo.Ctx, c.repo.Path, c.ID.String())
+	return CommitsCount(c.repo.Ctx, CommitsCountOptions{
+		RepoPath: c.repo.Path,
+		Not:      "",
```
Comment (lunny) [inline modules/git/commit.go]: blank parameter could be ignored.

File: modules/git/commit.go (hunk 2 above)
Comment (mattwalo32) [reply modules/git/commit.go]: Yep, I'll get rid of `CommitsCountFiles` and just use `CommitsCount` everywhere

File: modules/git/commit.go (hunk 3 above)
Comment (mattwalo32) [reply modules/git/commit.go]: Ah, it's been a while since I've worked with Go. Forgot that you could do that :). I'll also remove in all the other places we pass `Not: ""`

File: modules/git/commit.go [hunk 4]
```diff
@@ -160,30 +160,42 @@ func AllCommitsCount(ctx context.Context, repoPath string, hidePRRefs bool, file
...
-	return CommitsCountFiles(ctx, repoPath, revision, []string{})
-}
-
 // CommitsCount returns number of total commits of until current revision.
 func (c *Commit) CommitsCount() (int64, error) {
-	return CommitsCount(c.repo.Ctx, c.repo.Path, c.ID.String())
+	return CommitsCount(c.repo.Ctx, CommitsCountOptions{
+		RepoPath: c.repo.Path,
+		Revision: []string{c.ID.String()},
+		RelPath:  []string{},
```
Comment (lunny) [inline modules/git/commit.go]: This line could be removed.

File: modules/git/commit_test.go [hunk 5]
```diff
@@ -14,11 +14,32 @@ import (
 func TestCommitsCount(t *testing.T) {
 	bareRepo1Path := filepath.Join(testReposDir, "repo1_bare")
 
-	commitsCount, err := CommitsCount(DefaultContext, bareRepo1Path, "8006ff9adbf0cb94da7dad9e537e53817f9fa5c0")
+	commitsCount, err := CommitsCount(DefaultContext,
+		CommitsCountOptions{
+			RepoPath: bareRepo1Path,
+			Revision: []string{"8006ff9adbf0cb94da7dad9e537e53817f9fa5c0"},
+			RelPath:  []string{},
```
Comment (lunny) [inline modules/git/commit_test.go]: This line could be removed.

File: modules/git/commit_test.go [hunk 6]
```diff
@@ -14,11 +14,32 @@ import (
...
 
+func TestCommitsCountWithoutBase(t *testing.T) {
+	bareRepo1Path := filepath.Join(testReposDir, "repo1_bare")
+
+	commitsCount, err := CommitsCount(DefaultContext,
+		CommitsCountOptions{
+			RepoPath: bareRepo1Path,
+			Not:      "master",
+			Revision: []string{"branch1"},
+			RelPath:  []string{},
```
Comment (lunny) [inline modules/git/commit_test.go]: This line could be removed.

File: modules/git/repo_commit.go [hunk 7]
```diff
@@ -365,11 +383,20 @@ func (repo *Repository) CommitsBetweenIDs(last, before string) ([]*Commit, error
 
 // CommitsCountBetween return numbers of commits between two commits
 func (repo *Repository) CommitsCountBetween(start, end string) (int64, error) {
-	count, err := CommitsCountFiles(repo.Ctx, repo.Path, []string{start + ".." + end}, []string{})
+	count, err := CommitsCount(repo.Ctx, CommitsCountOptions{
+		RepoPath: repo.Path,
+		Revision: []string{start + ".." + end},
+		RelPath:  []string{},
```
Comment (lunny) [inline modules/git/repo_commit.go]: As above.

File: modules/git/repo_commit.go [hunk 8]
```diff
@@ -365,11 +383,20 @@ func (repo *Repository) CommitsBetweenIDs(last, before string) ([]*Commit, error
...
+	})
+
 	if err != nil && strings.Contains(err.Error(), "no merge base") {
 		// future versions of git >= 2.28 are likely to return an error if before and last have become unrelated.
 		// previously it would return the results of git rev-list before last so let's try that...
-		return CommitsCountFiles(repo.Ctx, repo.Path, []string{start, end}, []string{})
+		return CommitsCount(repo.Ctx, CommitsCountOptions{
+			RepoPath: repo.Path,
+			Revision: []string{start, end},
+			RelPath:  []string{},
```
Comment (lunny) [inline modules/git/repo_commit.go]: As above.

File: routers/api/v1/repo/commits.go [hunk 9]
```diff
@@ -178,14 +179,19 @@ func GetAllCommits(ctx *context.APIContext) {
 		}
 
 		// Total commit count
-		commitsCountTotal, err = baseCommit.CommitsCount()
+		commitsCountTotal, err = git.CommitsCount(ctx.Repo.GitRepo.Ctx, git.CommitsCountOptions{
+			RepoPath: ctx.Repo.GitRepo.Path,
+			Not:      not,
+			Revision: []string{baseCommit.ID.String()},
+			RelPath:  []string{},
```
Comment (lunny) [inline routers/api/v1/repo/commits.go]: As above

File: services/repository/cache.go [hunk 10]
```diff
@@ -34,7 +34,11 @@ func CacheRef(ctx context.Context, repo *repo_model.Repository, gitRepo *git.Rep
 	}
 
 	if gitRepo.LastCommitCache == nil {
-		commitsCount, err := cache.GetInt64(repo.GetCommitsCountCacheKey(getRefName(fullRefName), true), commit.CommitsCount)
+		getCount := func() (int64, error) {
+			return commit.CommitsCount()
+		}
+
+		commitsCount, err := cache.GetInt64(repo.GetCommitsCountCacheKey(getRefName(fullRefName), true), getCount)
```
Comment (lunny) [inline services/repository/cache.go]:
```suggestion
		commitsCount, err := cache.GetInt64(repo.GetCommitsCountCacheKey(getRefName(fullRefName), true), commit.CommitsCount)
```

**Review by lunny — APPROVED** [review]

File: modules/git/commit.go [hunk 11]
```diff
@@ -160,30 +160,41 @@ func AllCommitsCount(ctx context.Context, repoPath string, hidePRRefs bool, file
...
+
+	cmd.AddDynamicArguments(opts.Revision...)
+	if len(opts.RelPath) > 0 {
+		cmd.AddDashesAndList(opts.RelPath...)
 	}
 
-	stdout, _, err := cmd.RunStdString(&RunOpts{Dir: repoPath})
+	if opts.Not != "" {
+		cmd.AddOptionValues("--not", opts.Not)
+	}
```
Comment (wxiaoguang) [inline modules/git/commit.go]: It seems not right.

option/value should be before dashes/list

valid: `git --not xxx -- file1 file2`

invalid: `git -- file file2 --not xxx`

File: modules/git/repo_commit.go [hunk 12]
```diff
@@ -204,13 +204,26 @@ func (repo *Repository) FileChangedBetweenCommits(filename, id1, id2 string) (bo
...
-	return CommitsCountFiles(repo.Ctx, repo.Path, []string{revision}, []string{file})
+func (repo *Repository) FileCommitsCount(revision, file, not string) (int64, error) {
+	return CommitsCount(repo.Ctx,
+		CommitsCountOptions{
+			RepoPath: repo.Path,
+			Not:      not,
+			Revision: []string{revision},
+			RelPath:  []string{file},
+		})
+}
```
Comment (wxiaoguang) [inline modules/git/repo_commit.go]: There is only one call to `FileCommitsCount(..., not)`, other calls are all `FileCommitsCount(..., "")`

I think that only `FileCommitsCount(..., not)` could be changed to `CommitsCount(repo.Ctx, ...)`, then the other `FileCommitsCount(..., "")` could be reverted to ``FileCommitsCount(...)` to avoid unnecessary changes and make them clear.

Otherwise overall L-G-T-M.

File: modules/git/commit.go [hunk 13]
```diff
@@ -160,30 +160,42 @@ func AllCommitsCount(ctx context.Context, repoPath string, hidePRRefs bool, file
...
+func CommitsCount(ctx context.Context, opts CommitsCountOptions) (int64, error) {
 	cmd := NewCommand(ctx, "rev-list", "--count")
-	cmd.AddDynamicArguments(revision...)
-	if len(relpath) > 0 {
-		cmd.AddDashesAndList(relpath...)
+
+	cmd.AddDynamicArguments(opts.Revision...)
+
+	if opts.Not != "" {
+		cmd.AddOptionValues("--not", opts.Not)
```
Comment (wxiaoguang) [inline modules/git/commit.go:178]: Sorry for bothering ... the last nit (hopefully)

It's better to put  `AddOptionValues`  before `AddDynamicArguments` to avoid breaking git command

The widely used order is:  `option-values` -> `non-option dynamic arguments` -> `dashes + file list`

File: modules/git/repo_commit.go [hunk 14]
```diff
@@ -220,10 +232,15 @@ func (repo *Repository) CommitsByFileAndRange(revision, file string, page int) (
...
 		gitCmd := NewCommand(repo.Ctx, "rev-list").
-			AddOptionFormat("--max-count=%d", setting.Git.CommitsRangeSize*page).
+			AddOptionFormat("--max-count=%d", setting.Git.CommitsRangeSize*opts.Page).
 			AddOptionFormat("--skip=%d", skip)
-		gitCmd.AddDynamicArguments(revision)
-		gitCmd.AddDashesAndList(file)
+		gitCmd.AddDynamicArguments(opts.Revision)
+
+		if opts.Not != "" {
+			gitCmd.AddOptionValues("--not", opts.Not)
```
Comment (wxiaoguang) [inline modules/git/repo_commit.go:240]: ~~(and here, as above, put AddOptionValues before  AddDynamicArguments)~~

File: modules/git/commit.go (hunk 13 above)
Comment (mattwalo32) [reply modules/git/commit.go:178]: I've found that the command does not work as expected when adding  `--not` before the revision.

For example, if I make a new branch on a repo off master, named `branch1`. Then add 1 commit

```
git rev-list --not master --count branch1
```
Always returns 0

But
```
git rev-list --count branch1 --not master
```

Returns 1 as expected

Comment (mattwalo32) [reply modules/git/commit.go:178]: You can even see in the documentation here, under the `Description` section, they show an example where `--not` is placed after the revision

https://git-scm.com/docs/git-rev-list

```
git rev-list A B --not $(git merge-base --all A B)
```

Comment (wxiaoguang) [reply modules/git/commit.go:178]: Oh, you are right, thank you ~~~

**Review by wxiaoguang — APPROVED** [review]
